[Tooling] Do some `Fastfile` cleanup

AliSoftware opened the pull request:
Just some cleanup and maintenance in preparation for some more upcoming automation work on the `Fastfile` soon

- Remove extra instructions from Slack message—because those instructions are now in the MC scenario and we prefer to only have one source of truth
- Remove `after_confirming_push` helper—because this will become obsolete once we migrate to the lanes being run by CI
- Clean up and standardize YARD doc for lanes

AliSoftware committed:
Remove extra instructions from Slack message

Now that those instructions are in the MC scenario, better only keep a single source of truth for those.

AliSoftware committed:
Remove `after_confirming_push` helper

We won't need it when we'll move to CI anyway so makes the code simpler to read

AliSoftware committed:
Clean up and standardize YARD doc for lanes

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1105,9 +1117,13 @@ inal copy, and try again.")
-  # - Loop though all PRs which were still open and were assigned the milestones of the code-frozen version,
-  #   and move them to the next milestone + leaving a PR comment about it
+  # Update the milestone of still-open PRs
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/7cbbdbbc913b3a3c47ffe9e3209e81a16981863c/fastlane/Fastfile#L1121"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

AliSoftware committed:
Fix Rubocop violations

leandroalonso commented on fastlane/Fastfile:
```diff
@@ -306,15 +314,15 @@ platform :ios do
-      set_branch_protection(repository: GHHELPER_REPO, branch: release_branch_name)
-    end
+    trigger_beta_build(branch_to_build: release_branch_name)
+    create_backmerge_pr
```
@AliSoftware I left a question on Slack about this one.

I know this is going in the direction of running everything to a machine but:

1. How it is now, it will open the back merge PR even if the beta build fails for any reason. (while before it would only open the PR if the build succeeded)
2. [PRs are not created](https://github.com/Automattic/pocket-casts-ios/pulls?q=is%3Apr+author%3Apocketcasts+is%3Aclosed) by our pocketcasts user anymore
3. Since the PR is created by whoever ran the command, someone else from the team needs to approve. This is a small friction but the process was super lean before. Most of the time the release manager would approve it.

I believe moving it to a machine will reinstate the `pocketcasts` user as the creator of those PRs but creating it even after the build failed seems a bit wrong.

AliSoftware replied:
Yes the fact that the PR is not created by the bot will be fixed once I make those lanes be run by CI (but I got delayed a bit on that work, thought I'd have happened sooner, sorry)

As for creating the backmerge PR before the build has finished, I did this change intentionally (via https://github.com/Automattic/pocket-casts-ios/commit/b3fa8e7fc8a1522dfdc18915257b29f0bcca6171) back then.

My rationale for it was to:
 - Create the backmerge PR ASAP, so that the Release Manager would not have to wait for the beta or release build to finish building before the backmerge PR is created only at the end of the build… which would then itself trigger its own set of CI checks (the classical ones for any PR, like unit tests etc), and thus require you to wait even more for that CI build to finish too before you can merge the backmerge PR.
 - Instead, now the 2 CI builds (build for the beta/release and CI checks for the backmerge PR) are run in parallel, leading you to less wait time and ability to review and approve the backmerge PR sooner[^1]
 - In theory at least, I wouldn't expect the beta build to fail to build, because everything that lands in the `release/*` branch did that via a dedicated PR already, so whatever is in `release/*` and being built by the beta build in CI is supposed to have already been proven to build successfully when it was validated in the PR landing the beta fix(es).
 - The only exception I could think of for this if the CI build for the beta fail not due to a compilation error but due to e.g. an ASC API error, or some issue when uploading the build to TestFlight. But those should be rare and in most cases either you'd not need to add any new commit to fix it (e.g. just a retry or similar), or you'd retrigger a `new_beta_release` after having pushed a commit to fix the issue, and that will close the previous backmerge PR and create a new one.

[^1]: All of this setup was assuming that we'd have moved to the lanes being run by CI sooner, and thus that those backmerge PRs would be created by the bot that runs the CI, allowing the Release Manager to approve and merge them themselves. This is still the plan, it just won't happen until January do to unforeseen delays on my project for this. But of course I understand that the fact that currently that's not yet the case and the PR is currently authored by the Release Manager instead of the bot kinda throws a spanner in the works of this reasoning until it will happen.

---

That being said, I'm OK if you prefer that we revert to do the backmerge PR only at the end of the CI build, even if this means the process will be back to be sequential and will require you to wait longer before you can start reviewing the PR.
But I just wanted to highlight my initial rationale in case it helped making sense of the initial plans 😉 

_PS: Also if we end up deciding to revert to doing the backmerge PR only after the build has finished, I'll probably do that change in a follow-up PR rather than this one, to keep the diff easier to review and on topic._

leandroalonso replied:
> I'll probably do that change in a follow-up PR rather than this one, to keep the diff easier to review and on topic.

Oh, absolutely! Don't worry!

I think we're aligned in terms of speeding up the waiting time. 🤓 My main concern with the PR being open by a member it's that the release manager itself cannot approve. So if they want to move quickly that might require pinging someone else. (That's what happened today, by the way)

> Create the backmerge PR ASAP, so that the Release Manager would not have to wait for the beta or release build to finish building before the backmerge PR is created only at the end of the build… which would then itself trigger its own set of CI checks (the classical ones for any PR, like unit tests etc), and thus require you to wait even more for that CI build to finish too before you can merge the backmerge PR.

Regarding that, I just want to mention that my workflow wasn't affected by the CI time because I would just approve and enable auto-merge. Then I'd just move on.

I think your rationale is good, the only issue is that one of us opening the PR requires someone else to approve, so a lot of time it will require a ping to someone else.

> The only exception I could think of for this if the CI build for the beta fail not due to a compilation error but due to e.g. an ASC API error, or some issue when uploading the build to TestFlight. But those should be rare 

That happened in the past. 😅 And I believe more recently we had the Sentry part failing.

Anyway, I think my only request would be to open the PR with the `pocketcasts` user, the rest is fine. But if we're doing that by January then it's fine to wait.

A temporary solution might be to use the Github API token from the `.mobile-secrets` to open the PRs, but I'll leave the decision to you.

Thanks for taking the time to explain!

AliSoftware committed:
Remove some obsolete lanes +  reorder for consistency

Now that the build and the upload are done by separate jobs on CI (both for Prototype Builds in `pipeline.yml` and for Beta/Final builds in `release-builds.yml`), we don't need the previous lanes that did both in one go.

IIRC back when we split the lanes in 2 jobs it was argued that it could be useful to keep the old lanes around in case one would need to run them locally.
But I'd argue that this makes us keep some dead/legacy code around that won't be maintained, and is likely to go stale, and clutters the Fastfile and makes it less easy to navigate and reason about, so I think it's clearer if we remove this unused code after all.

iangmaia commented on fastlane/Fastfile:
```diff
@@ -306,15 +314,15 @@ platform :ios do
+
+    # We cannot use the `copy_branch_protection` action here yet, as it would create a branch protection rule with the same
+    # restrictions we have in `trunk`, and the release managers wouldn't be able to push due to permissions.
+    # This should be changed only when we have PCiOS releases done on CI, when the CI bot is the one running `git push`.
```
👍

iangmaia reviewed (approved):
Nice improvements and clean-up 🧹
